### PR TITLE
feat: proper types for rows (#46)

### PR DESCRIPTION
**Pull request by spike-rabbit**
* feat: use proper type for row related properties

* fix: detect update in rows automatically if row identity changed

* perf: only detect changes in rows on real changes

* docs: use row related types in examples

**What kind of change does this PR introduce?** (check one with "x")

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")

- [x] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: projects/ngx-datatable/src/lib/components/header/header.component.ts [hunk 1]
```diff
@@ -150,25 +155,25 @@ export class DataTableHeaderComponent implements OnDestroy, OnChanges {
...
   _headerHeight: string;
-  _styleByGroup: { [prop: string]: unknown } = {
-    left: {},
-    center: {},
-    right: {}
+  _styleByGroup = {
+    left: NgStyle['ngStyle'],
+    center: NgStyle['ngStyle'],
+    right: NgStyle['ngStyle']
   };
```
Comment (chintankavathia) [inline projects/ngx-datatable/src/lib/components/header/header.component.ts:177]:
```suggestion
_styleByGroup: {
    left: NgStyle['ngStyle'],
    center: NgStyle['ngStyle'],
    right: NgStyle['ngStyle']
  } = {
      left: {},
      center: {},
      right: {}
    };
```

File: projects/ngx-datatable/src/lib/directives/long-press.directive.ts [hunk 2]
```diff
@@ -2,16 +2,18 @@ import { Directive, EventEmitter, HostBinding, HostListener, Input, OnDestroy, O
...
 @Directive({ selector: '[long-press]' })
 export class LongPressDirective implements OnDestroy {
   @Input() pressEnabled = true;
-  @Input() pressModel: any;
+  @Input() pressModel: TableColumn;
   @Input() duration = 500;
 
-  @Output() longPressStart: EventEmitter<any> = new EventEmitter();
+  @Output() longPressStart: EventEmitter<{event: MouseEvent; model: TableColumn}> = new EventEmitter();
   @Output() longPressing: EventEmitter<any> = new EventEmitter();
```
Comment (chintankavathia) [inline projects/ngx-datatable/src/lib/directives/long-press.directive.ts:15]: I guess we can type this as well

File: projects/ngx-datatable/src/lib/types/activate-event.type.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  type: 'checkbox' |'click' | 'dblclick' | 'keydown';
+  event: MouseEvent | KeyboardEvent;
+  row: RowOrGroup<TRow>;
+  group: TRow[];
+  rowHeight: number;
+  column: TableColumn;
+  value: any;
+  cellElement: HTMLElement;
+  treeStatus?: TreeStatus;
+
```
Comment (chintankavathia) [inline projects/ngx-datatable/src/lib/types/activate-event.type.ts:15]:
```suggestion
```

File: projects/ngx-datatable/src/public-api.ts [hunk 4]
```diff
@@ -52,6 +52,7 @@ export * from './lib/types/click.type';
 export * from './lib/types/table-column.type';
 export * from './lib/types/sort-prop-dir.type';
 export * from './lib/types/contextmenu.type';
+export * from './lib/types/orderable.types';
```
Comment (chintankavathia) [inline projects/ngx-datatable/src/public-api.ts:55]: shouldn't we export other types as well?

**Review by chintankavathia — CHANGES_REQUESTED** [review]

**Comment by chintankavathia** [comment]
@spike-rabbit I left some comments.

File: projects/ngx-datatable/src/lib/components/body/body-row.component.ts [hunk 5]
```diff
@@ -148,25 +151,25 @@ export class DataTableBodyRowComponent implements DoCheck, OnChanges {
   @Output() activate: EventEmitter<any> = new EventEmitter();
   @Output() treeAction: EventEmitter<any> = new EventEmitter();
 
-  _element: any;
+  _element: HTMLElement;
   _columnGroupWidths: any;
```
Comment (chintankavathia) [inline projects/ngx-datatable/src/lib/components/body/body-row.component.ts:155]: {left: number, center: number, right: number, total: number}

File: projects/ngx-datatable/src/lib/components/body/body.component.ts [hunk 6]
```diff
@@ -208,7 +217,8 @@ export class DataTableBodyComponent implements OnInit, OnDestroy {
     this._ghostLoadingIndicator = val;
     if (!val) {
       // remove placeholder rows once ghostloading is set to false
-      this.temp = this.temp.filter(item => !!item);
+      // typescript is cannot detect the type --> casting
```
Comment (chintankavathia) [inline projects/ngx-datatable/src/lib/components/body/body.component.ts:220]:
```suggestion
      // typescript cannot detect the type --> casting
```

File: projects/ngx-datatable/src/lib/components/body/body.component.ts [hunk 7]
```diff
@@ -670,7 +680,7 @@ export class DataTableBodyComponent implements OnInit, OnDestroy {
     if (this.scrollbarV && this.virtualization) {
       let idx = 0;
 
-      if (this.groupedRows) {
+      if (rows instanceof Array) {
```
Comment (chintankavathia) [inline projects/ngx-datatable/src/lib/components/body/body.component.ts:683]:
```suggestion
      if (Array.isArray(rows)) {
```

File: projects/ngx-datatable/src/lib/components/datatable.component.ts [hunk 8]
```diff
@@ -366,13 +374,11 @@ export class DatatableComponent implements OnInit, DoCheck, AfterViewInit, After
...
-   * Similar implementation to ngClass.
-   *
-   *  [rowClass]="'first second'"
-   *  [rowClass]="{ 'first': true, 'second': true, 'third': false }"
+   * A function which is called with the row and should return either:
+   * - a string: `"class-1 class-2`
+   * - a Record<string, boolean>: `{ 'class-1': true, 'class-2': false }`
    */
-  @Input() rowClass: any;
+  @Input() rowClass: (row: Group<TRow> | TRow) => string | Record<string, boolean>;
```
Comment (chintankavathia) [inline projects/ngx-datatable/src/lib/components/datatable.component.ts:381]: I think we shouldn't just limit it to be a function but also allow users to pass `string | Record<string, boolean>` directly.

File: projects/ngx-datatable/src/lib/components/datatable.component.ts [hunk 9]
```diff
@@ -823,14 +829,15 @@ export class DatatableComponent implements OnInit, DoCheck, AfterViewInit, After
    * Creates a map with the data grouped by the user choice of grouping index
    *
    * @param originalArray the original array passed via parameter
-   * @param groupBy  the index of the column to group the data by
+   * @param groupBy  the key of the column to group the data by
```
Comment (chintankavathia) [inline projects/ngx-datatable/src/lib/components/datatable.component.ts:832]:
```suggestion
   * @param groupBy the key of the column to group the data by
```

**Review by chintankavathia — CHANGES_REQUESTED** [review]

File: projects/ngx-datatable/src/lib/components/body/body.component.ts [hunk 10]
```diff
@@ -221,23 +231,23 @@ export class DataTableBodyComponent implements OnInit, OnDestroy {
   @Input() selectionType: SelectionType;
   @Input() selected: any[] = [];
```
[the remark this answers is not in the document]
Comment (spike-rabbit) [reply projects/ngx-datatable/src/lib/components/body/body.component.ts:232]: We have to postpone this. Changing this gives a type error, as TRow of Body may not be the same as from the selection component. I will fix this later.

File: projects/ngx-datatable/src/lib/components/datatable.component.ts (hunk 8 above)
Comment (spike-rabbit) [reply projects/ngx-datatable/src/lib/components/datatable.component.ts:381]: you are right in general. But the implementation currently only supports this function. We can do this in another MR.

**Comment by chintankavathia** [comment]
@spike-rabbit Not sure but shouldn't we add a breaking change note here? otherwise LGTM

**Comment by chintankavathia** [comment]
> breaking change added

There is one conflict which needs to be resolved locally.

File: projects/ngx-datatable/src/lib/components/datatable.component.ts [hunk 11]
```diff
@@ -655,7 +662,7 @@ export class DatatableComponent implements OnInit, DoCheck, AfterViewInit, After
    * invoking functions on the body.
    */
   @ViewChild(DataTableBodyComponent)
-    bodyComponent: DataTableBodyComponent;
+    bodyComponent: DataTableBodyComponent<TRow>;
```
Comment (chintankavathia) [inline projects/ngx-datatable/src/lib/components/datatable.component.ts:665]:
```suggestion
    bodyComponent: DataTableBodyComponent<TRow & { treeStatus?: TreeStatus }>;
```

**Review by timowolf — APPROVED** [review]
Works like a charm. Not all types are covered and some internal types are exported. As discussed, we will block master branch for the moment until we have all big changes concerning types, ng18 and names right.

I tested in multiple applications and did not find any problem.

**Pull request merged**